crypto/evp/evp_utils.c

```diff
@@ -17,31 +17,66 @@
 #include "internal/evp_int.h"    /* evp_locl.h needs it */
 #include "evp_locl.h"
 
-int evp_do_ciph_getparams(const EVP_CIPHER *ciph, OSSL_PARAM params[])
-{
-    if (ciph->prov == NULL)
-        return -2;
-    if (ciph->get_params == NULL)
-        return -1;
-    return ciph->get_params(params);
+
+/*
+ * EVP_CTRL_RET_UNSUPPORTED = -1 is the returned value from any ctrl function
+ * where the control command isn't supported, and an alternative code path
+ * may be chosen.
+ * Since these functions are used to implement ctrl functionality, we
+ * use the same value, and other callers will have to compensate.
+ */
+#define EVP_PARAM_CHECK(obj, func, errfunc)                                    \
+    if (obj->prov == NULL)                                                     \
+        return EVP_CTRL_RET_UNSUPPORTED;                                                    \
+    if (obj->func == NULL) {                                                   \
+        errfunc();                                                             \
+        return 0;                                                              \
+    }
+
+#define EVP_PARAM_FUNC(name, func, type, err)                                  \
+int name (const type *obj, OSSL_PARAM params[])                                \
+{                                                                              \
+    EVP_PARAM_CHECK(obj, func, err)                                            \
+    return obj->func(params);                                                  \
+}
+
+#define EVP_PARAM_CTX_FUNC(name, func, type, err)                              \
+int name (const type *obj, void *provctx, OSSL_PARAM params[])                 \
+{                                                                              \
+    EVP_PARAM_CHECK(obj, func, err)                                            \
+    return obj->func(provctx, params);                                         \
 }
 
-int evp_do_ciph_ctx_getparams(const EVP_CIPHER *ciph, void *provctx,
-                              OSSL_PARAM params[])
+#define EVP_PARAM_FUNCTIONS(type,                                              \
+                            getname, getfunc,                                  \
+                            getctxname, getctxfunc,                            \
+                            setctxname, setctxfunc)                            \
+    EVP_PARAM_FUNC(getname, getfunc, type, geterr)                             \
+    EVP_PARAM_CTX_FUNC(getctxname, getctxfunc, type, geterr)                   \
+    EVP_PARAM_CTX_FUNC(setctxname, setctxfunc, type, seterr)
+
+/*
+ * These error functions are a workaround for the error scripts, which
+ * currently require that XXXerr() appears inside a function (not a macro).
```

Comment:
And apparently not inside a comment either, see the Travis failure.
I think that if you lose the (), then mkerr.pl might not see it...

Reply:
Doesn't this mean that the Perl script should be fixed instead 😈